File: #199_Binary_Tree_Right_Side_View/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new LinkedList<Integer>();
        if (root == null) return result;
        Deque<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        int cur = 1;
        int next = 0;
        while (!queue.isEmpty()) {
        	TreeNode node = queue.remove();
        	// System.out.println("pop: " + node.val);
        	cur--;
        	if (node.left != null) {
        		queue.add(node.left);
        		next++;
        	}
        	if (node.right != null) {
        		queue.add(node.right);
        		next++;
        	}
        	if (cur == 0) {
        		result.add(node.val);
        		cur = next;
        		next = 0;
        	}
        }
        return result;
    }

    public static void main(String[] args) {
    	TreeNode head = new TreeNode(1);
    	TreeNode n1 = new TreeNode(2);
    	TreeNode n2 = new TreeNode(3);
    	head.left = n1;
    	head.right = n2;
    	Solution sol = new Solution();
    	List<Integer> result = sol.rightSideView(head);
    	for (Integer it : result) {
    		System.out.println(it);
    	}
    }
}

File: #199_Binary_Tree_Right_Side_View/Solution.py
from collections import deque

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def rightSideView(self, root):
    	if not root:
    		return []
    	queue = deque([root])
    	re = []
    	cur, next = 1, 0
    	while queue:
    		node = None
    		for x in range(cur):
    			node = queue.popleft()
    			if node.left:
    				queue.append(node.left)
    				next += 1
    			if node.right:
    				queue.append(node.right)
    				next += 1
    		re.append(node.val)
    		cur = next
    		next = 0
    	return re

sol = Solution()
root = TreeNode(10)
root.left = TreeNode(1)
root.right = TreeNode(20)
root.left.left = TreeNode(-3)
# root.right.right = TreeNode(50)
print sol.rightSideView(root)